library/aes.c

```diff
@@ -1441,36 +1442,38 @@ int mbedtls_aes_crypt_ctr(mbedtls_aes_context *ctx,
                           const unsigned char *input,
                           unsigned char *output)
 {
-    int c, i;
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    size_t n;
 
-    n = *nc_off;
+    size_t offset = *nc_off;
 
-    if (n > 0x0F) {
+    if (offset > 0x0F) {
         return MBEDTLS_ERR_AES_BAD_INPUT_DATA;
     }
 
-    while (length--) {
-        if (n == 0) {
+    for (size_t i = 0; i < length;) {
+        size_t n = 16;
+        if (offset == 0) {
             ret = mbedtls_aes_crypt_ecb(ctx, MBEDTLS_AES_ENCRYPT, nonce_counter, stream_block);
             if (ret != 0) {
                 goto exit;
             }
-
-            for (i = 16; i > 0; i--) {
-                if (++nonce_counter[i - 1] != 0) {
```

Comment:
Also in aria.c:673 and camellia.c:644 and ccm.c:455

Similar at ctr_drbg.c:349 and 658

A 4-byte one at pkcs5.c:343

Similar at ssl_msg.c:3040,4057 ssl_tls.c:8039

Reply:
I haven't done aria and camellia because we don't have benchmarks here, and they are less widely used than AES.

For CCM, the number of iterations is not constant, so this isn't a quick drop-in change - probably it can still work but it's less of a quick win.

For PKCS and SSL, I don't think it's in a performance sensitive bit, and also in some cases, the number of iterations is again not constant.

Reply:
I would naively expect Camellia and ARIA to benefit in similar amounts since they are similar to AES (when no AES-specific instructions are involved). That doesn't mean we have to optimize them, since indeed there is less demand. Note that for ARIA, we'd want to have more extensive testing (there are currently no unit tests for multipart ARIA-CTR).

CCM would likely be a similar win as GCM and CTR, but maybe not since it can depend on how the counter calculations are mixed with the rest.

PKCS5 is practically only used to decrypt keys on non-highly-constrained platforms so not worth the comparatively very minor optimization.

The uses in TLS are all 8 bytes or less, so I wonder if a better approach might be to store and increment the value as a `uint64_t`, and format it with the appropriate endianness?

Reply:
On a 64-bit machine, for the CTR cases, using 64-bit reads/writes is not measurably different to the 32-bit implementation (and I guess it will perform same-or-worse on 32-bit machines).

For TLS, if it's a single 64-bit read/increment/write, then it makes sense to express using `uint64_t`. But I suspect that this will not result in a measurable overall difference, given network latencies.

There's definitely more that can be done, but I don't want the scope of this PR to grow indefinitely so I'm saying no to all the above for now and making a mental note to look at ccm when I have a bit of spare time.